### DISCUSSION TITLE
incrementally maintained views

### DISCUSSION THREAD

**Discussion by docteurklein** [cozodb/cozo#35] [Ideas]
Hello!

How awesome would it be to have https://github.com/TimelyDataflow/differential-dataflow used to maintain materialized views? 
Something like https://materialize.com/ or https://github.com/vmware/differential-datalog.

Just a random idea but wanted to have your opinions :). Thanks for the great work!

**Comment by zh217** [comment]
It would certainly be great, but at the moment I am not too sure how to proceed with this feature.

Do we want incremental views for arbitrary queries? This is an active area of research, there are some results, but this problem is actually unsolvable in the general case (in the sense that lots of queries are not amenable to incremental views---attempts to do so are actually slower than recomputing everything from scratch on every update). So we can't really allow arbitrary queries, otherwise casual use of this feature may have extremely adverse effects on performance.

What about restricting the queries to the dataflow-y subset? The first problem is to find out what exactly is this subset, which I haven't done yet (since we allow arbitrary computation in the evaluation of expressions which may later be joined, this is a difficult question even if we disallow all recursions). And then the query engine needs a separate implementation for the streaming case (I don't think semi-naive is still a good fit anymore). The long-term maintainability of two separate engines is questionable at this stage.

So at the moment there is no plan to add this functionality. What is needed is a much clearer _theoretical_ understanding of the problems involved than we currently have now.

**Reply by docteurklein** [reply]
Thanks! I agree that implementing an engine on top of data flow seems like a gigantic pile of work, and surely requires careful thinking around recursive rules and such (even materialize.com don't support some queries.

Another approach, taken by pg_ivm and others is to use "statement triggers" (i.e cozo's `new[]` relations in triggers) to maintain a view (implementable subsets at least).
It'd be great to infer triggers from a view definition.
